File: code_samples/section11/problems/problem11_2/problem11_2.cs
static void DfsComp(int u, int cid, List<List<int>> adj, int[] comp) {
    // Depth-First Search (DFS) helper that labels all nodes reachable from `u`
    // with the connected-component id `cid`.
    //
    // Parameters:
    //   u    : current node being visited
    //   cid  : component id to assign to all nodes in this connected component
    //   adj  : adjacency list for an UNDIRECTED graph; adj[u] lists neighbors of u
    //   comp : output array; comp[x] == -1 means "unvisited", otherwise stores component id
    //
    // Behavior:
    //   - Marks comp[u] = cid
    //   - Recursively visits every neighbor v that has not yet been labeled
    //
    // Note:
    //   - This is recursive DFS; extremely large/deep graphs can cause stack overflow.
    comp[u] = cid;

    foreach (var v in adj[u]) {
        // If v is unvisited, it belongs to the same component.
        if (comp[v] == -1) {
            DfsComp(v, cid, adj, comp);
        }
    }
}

static int ConnectedComponents(int n, List<List<int>> adj, int[] comp) {
    // Computes connected components in an UNDIRECTED graph using DFS labeling.
    //
    // Parameters:
    //   n    : number of nodes (assumed labeled 0..n-1)
    //   adj  : adjacency list
    //   comp : caller-provided output array of length n; will be overwritten
    //
    // Returns:
    //   cid  : the number of connected components discovered
    //
    // Approach:
    //   - Initialize comp[] to -1 (meaning "unvisited")
    //   - For each node u:
    //       - if unvisited, start DFS from u and label the entire component with cid
    //       - increment cid after finishing that component
    Array.Fill(comp, -1);

    int cid = 0;

    for (int u = 0; u < n; u++) {
        if (comp[u] == -1) {
            DfsComp(u, cid, adj, comp);
            cid++;
        }
    }

    // When done, cid equals the total number of components.
    return cid;
}

// ===== Helper Functions =====

List<List<int>> MakeGraph(int n)
{
    // Allocates an empty adjacency list for a graph with n nodes.
    //
    // Returns:
    //   g : List<List<int>> where g[i] is the (initially empty) neighbor list for node i
    //
    // Note:
    //   - This sets up structure only; edges are added with AddEdge().
    var g = new List<List<int>>(n);

    // Pre-populate so g[i] is valid for every node.
    for (int i = 0; i < n; i++)
        g.Add([]); // target-typed empty list (modern C# collection expression)
    return g;
}

void AddEdge(List<List<int>> adj, int u, int v)
{
    // Adds an UNDIRECTED edge between u and v.
    //
    // Parameters:
    //   adj : adjacency list
    //   u,v : endpoints (assumed valid indices)
    //
    // Behavior:
    //   - Adds v to adj[u] and u to adj[v].
    //   - Parallel edges / self-loops are allowed if provided.
    adj[u].Add(v);
    adj[v].Add(u);
}

void RunTest(string name, int n, List<List<int>> adj)
{
    // Runs the connected-components algorithm on the provided graph and prints:
    //   - the number of components
    //   - the component id assigned to each node
    //
    // Parameters:
    //   name : label printed for the test case
    //   n    : number of nodes
    //   adj  : adjacency list
    //
    // Note:
    //   - This is a "print and inspect" harness; it does not assert correctness.
    Console.WriteLine(name);

    // Allocate comp[] for this run; ConnectedComponents will fill it.
    int[] comp = new int[n];

    int count = ConnectedComponents(n, adj, comp);

    Console.WriteLine($"Component count = {count}");
    Console.WriteLine("Component IDs   = " + string.Join(", ", comp));
    Console.WriteLine();
}

// ===============================
// Test 1: Fully Connected Graph
// ===============================

{
    Console.WriteLine("=== Test 1: Fully connected graph ===");

    int n = 5;
    var adj = MakeGraph(n);

    // Build a clique (complete graph) by adding an edge between every pair (u, v), u < v.
    // A complete graph is always a single connected component.
    for (int u = 0; u < n; u++)
        for (int v = u + 1; v < n; v++)
            AddEdge(adj, u, v);

    RunTest("Complete graph", n, adj);
}

// ===============================
// Test 2: Two Components + Isolated Node
// ===============================

{
    Console.WriteLine("=== Test 2: Two components + isolated ===");

    // Graph intended:
    //   - component 0: {0,1,2}
    //   - component 1: {3,4}
    //   - component 2: {5} (isolated)
    // Total expected components: 3
    int n = 6;
    var adj = MakeGraph(n);

    // Component 0: 0-1-2
    AddEdge(adj, 0, 1);
    AddEdge(adj, 1, 2);

    // Component 1: 3-4
    AddEdge(adj, 3, 4);

    // Node 5 isolated (no edges added)

    RunTest("Two components plus isolated", n, adj);
}

// ===============================
// Test 3: All Nodes Isolated
// ===============================

{
    Console.WriteLine("=== Test 3: All nodes isolated ===");

    // No edges => each node is its own component.
    // With n=4, expected component count is 4.
    int n = 4;
    var adj = MakeGraph(n); // no edges

    RunTest("4 isolated nodes", n, adj);
}

// ===============================
// Test 4: Mixed Graph
// ===============================

{
    Console.WriteLine("=== Test 4: Mixed graph ===");

    // Graph intended:
    //   - component 0: {0,1,2}
    //   - component 1: {3,4,5}
    //   - component 2: {6} isolated
    // Total expected components: 3
    int n = 7;
    var adj = MakeGraph(n);

    // Component 0: 0-1-2
    AddEdge(adj, 0, 1);
    AddEdge(adj, 1, 2);

    // Component 1: 3-4-5
    AddEdge(adj, 3, 4);
    AddEdge(adj, 4, 5);

    // Node 6 isolated (no edges added)

    RunTest("Mixed graph", n, adj);
}

File: code_samples/section11/problems/problem11_2/problem11_2.java
package code_samples.section11.problems.problem11_2;

import java.util.*;

/*
 * problem11_2
 * -----------
 * This class demonstrates how to find / label connected components in an
 * UNDIRECTED graph using Depth-First Search (DFS).
 *
 * High-level idea:
 *   - Maintain an array comp[] where comp[u] is the component id for node u.
 *   - Initialize comp[] to -1 (meaning "unvisited").
 *   - For each node u:
 *       - if comp[u] == -1, start a DFS from u and label every reachable node
 *         with the current component id.
 *       - increment component id and continue scanning.
 *
 * Graph representation:
 *   - The graph is an adjacency list: List<List<Integer>>
 *   - Nodes are assumed labeled 0..n-1.
 *
 * Test harness:
 *   - Builds several graphs (clique, two components + isolated, isolated nodes, mixed)
 *   - Prints component count and the component id assigned to each node.
 */
public class problem11_2 {

    /*
     * dfsComp
     * -------
     * Recursive DFS helper that labels all nodes reachable from `u`
     * with a given component id `cid`.
     *
     * Parameters:
     *   u    : current node being visited
     *   cid  : component id to assign to all nodes in this connected component
     *   adj  : adjacency list for an UNDIRECTED graph (adj.get(u) lists neighbors)
     *   comp : component id array; comp[x] == -1 indicates "unvisited"
     *
     * Behavior:
     *   - Sets comp[u] = cid
     *   - Recursively visits any neighbor v that is still unvisited (comp[v] == -1)
     *
     * Note:
     *   - This implementation is recursive. On very large/deep graphs it could
     *     overflow the call stack; an explicit stack can be used instead if needed.
     */
    void dfsComp(int u, int cid, List<List<Integer>> adj, int[] comp) {
        comp[u] = cid; // mark u as belonging to component cid

        // Explore all neighbors of u.
        for (int v : adj.get(u)) {
            // If neighbor has not been labeled yet, visit it.
            if (comp[v] == -1) {
                dfsComp(v, cid, adj, comp);
            }
        }
    }

    /*
     * connectedComponents
     * -------------------
     * Labels and counts connected components in an UNDIRECTED graph.
     *
     * Parameters:
     *   n    : number of nodes
     *   adj  : adjacency list
     *   comp : output array of length n; will be overwritten with component ids
     *
     * Returns:
     *   cid  : number of connected components found
     *
     * Approach:
     *   - Fill comp[] with -1 to indicate all nodes are initially unvisited.
     *   - Scan nodes 0..n-1:
     *       - if a node is unvisited, start a DFS to label its entire component.
     *       - increment cid after finishing each component.
     */
    int connectedComponents(int n, List<List<Integer>> adj, int[] comp) {
        Arrays.fill(comp, -1); // mark all nodes as unvisited

        int cid = 0; // current component id; also ends as the component count

        for (int u = 0; u < n; u++) {
            // If u has not been assigned a component yet, it begins a new component.
            if (comp[u] == -1) {
                dfsComp(u, cid, adj, comp);
                cid++;
            }
        }
        return cid;
    }

    /*
     * main
     * ----
     * Runs the provided test cases.
     */
    public static void main(String[] args) {
        problem11_2 solver = new problem11_2();

        testFullyConnected(solver);
        testTwoComponents(solver);
        testIsolatedNodes(solver);
        testMixedGraph(solver);
    }

    // Helper: create adjacency list
    static List<List<Integer>> makeGraph(int n) {
        /*
         * Allocates an empty adjacency list for a graph with n nodes.
         *
         * Returns:
         *   g : List<List<Integer>> where each g.get(i) is an initially empty list
         *       of neighbors for node i.
         */
        List<List<Integer>> g = new ArrayList<>(n);
        for (int i = 0; i < n; i++) g.add(new ArrayList<>());
        return g;
    }

    // Helper: undirected edge
    static void addEdge(List<List<Integer>> adj, int u, int v) {
        /*
         * Adds an UNDIRECTED edge between u and v.
         *
         * Behavior:
         *   - Adds v to u's neighbor list
         *   - Adds u to v's neighbor list
         *
         * Notes:
         *   - No bounds checking is performed on u and v.
         *   - Parallel edges/self-loops are allowed if provided.
         */
        adj.get(u).add(v);
        adj.get(v).add(u);
    }

    static void runTest(String name, problem11_2 solver, int n, List<List<Integer>> adj) {
        /*
         * Runs connected component labeling on the provided graph and prints:
         *   - test name
         *   - component count
         *   - component id array
         *
         * Note:
         *   - This is a "print and inspect" test harness (no assertions).
         */
        System.out.println(name);

        int[] comp = new int[n];
        int count = solver.connectedComponents(n, adj, comp);

        System.out.println("Component count = " + count);
        System.out.println("Component IDs   = " + Arrays.toString(comp));
        System.out.println();
    }

    // ======================
    // Test cases
    // ======================

    static void testFullyConnected(problem11_2 solver) {
        /*
         * Test 1: Fully connected graph (clique).
         *
         * A clique on n nodes has exactly 1 connected component,
         * because every node is reachable from every other node.
         */
        int n = 5;
        List<List<Integer>> adj = makeGraph(n);

        // clique: add every edge (u, v) for u < v
        for (int u = 0; u < n; u++)
            for (int v = u + 1; v < n; v++)
                addEdge(adj, u, v);

        runTest("=== Test 1: Fully connected graph ===", solver, n, adj);
    }

    static void testTwoComponents(problem11_2 solver) {
        /*
         * Test 2: Two components + isolated node.
         *
         * Graph structure:
         *   - component A: 0-1-2
         *   - component B: 3-4
         *   - component C: {5} isolated
         *
         * Expected component count: 3.
         */
        int n = 6;
        List<List<Integer>> adj = makeGraph(n);

        // component A: 0-1-2
        addEdge(adj, 0, 1);
        addEdge(adj, 1, 2);

        // component B: 3-4
        addEdge(adj, 3, 4);

        // node 5 isolated (no edges added)

        runTest("=== Test 2: Two components + isolated ===", solver, n, adj);
    }

    static void testIsolatedNodes(problem11_2 solver) {
        /*
         * Test 3: All nodes isolated.
         *
         * No edges => each node is its own component.
         * For n = 4, expected component count: 4.
         */
        int n = 4;
        List<List<Integer>> adj = makeGraph(n); // no edges

        runTest("=== Test 3: All nodes isolated ===", solver, n, adj);
    }

    static void testMixedGraph(problem11_2 solver) {
        /*
         * Test 4: Mixed graph.
         *
         * Graph structure:
         *   - component A: 0-1-2
         *   - component B: 3-4-5
         *   - component C: {6} isolated
         *
         * Expected component count: 3.
         */
        int n = 7;
        List<List<Integer>> adj = makeGraph(n);

        // Component A: 0-1-2
        addEdge(adj, 0, 1);
        addEdge(adj, 1, 2);

        // Component B: 3-4-5
        addEdge(adj, 3, 4);
        addEdge(adj, 4, 5);

        // Node 6 isolated

        runTest("=== Test 4: Mixed graph ===", solver, n, adj);
    }
}
